include/deal.II/lac/petsc_snes.h

```diff
@@ -0,0 +1,477 @@
+   * set an optional NonlinearSolver::energy callback.
+   */
+  template <typename VectorType  = PETScWrappers::VectorBase,
+            typename PMatrixType = PETScWrappers::MatrixBase,
+            typename AMatrixType = PMatrixType>
+  class NonlinearSolver
+  {
+  public:
+    /**
+     * Type that holds real-valued numbers.
+     *
+     * Used to represent norms.
+     */
+    using real_type = PetscReal;
+
+    /**
+     * Constructor.
+     */
+    NonlinearSolver(const NonlinearSolverData &data     = NonlinearSolverData(),
+                    const MPI_Comm &           mpi_comm = PETSC_COMM_WORLD);
+
+    /**
+     * Destructor.
+     */
+    ~NonlinearSolver();
+
+    /**
+     * Return the PETSc SNES object.
+     */
+    SNES
+    petsc_snes();
+
+    /**
+     * Return the underlying MPI communicator.
+     */
+    MPI_Comm
+    get_mpi_communicator() const;
+
+    /**
+     * Reset the solver, it does not change the customization.
+     */
+    void
+    reinit();
+
+    /**
+     * Reset solver.
+     * Change customization according to `data`.
+     */
+    void
+    reinit(const NonlinearSolverData &data);
+
+    /**
+     * Set the preconditioning matrix only.
+     *
+     * When used with NonlinearSolver::setup_jacobian and
+     * NonlinearSolver::solve_for_jacobian_system, PETSc will approximate the
+     * linear system matrix-vector product using an internal matrix-free
+     * representation.
+     *
+     * When used with NonlinearSolver::jacobian PETSc will use the same matrix
+     * for both preconditioning and matrix-vector products.
+     */
+    void
+    reinit_matrix(PMatrixType &P);
+
+    /**
+     * Set both the linear system matrix and the preconditioning matrix
+     * that PETSc will use.
+     */
+    void
+    reinit_matrices(AMatrixType &A, PMatrixType &P);
+
+    /**
+     * Solve the nonlinear system of equations $F(x) = 0$.
+     *
+     * This function returns the number of iterations.
+     * The vector @p x must contain the initial guess.
+     * Upon returning, the @p x vector contains the solution.
+     */
+    unsigned int
+    solve(VectorType &x);
+
+    /**
+     * Solve the nonlinear system of equations $F(x) = 0$.
+     *
+     * This function returns the number of iterations.
+     * The vector @p x must contain the initial guess.
+     * Upon returning, the @p x vector contains the solution.
+     *
+     * Here we also set the matrix to handle the Jacobian.
+     */
+    unsigned int
+    solve(VectorType &x, PMatrixType &P);
+
+    /**
+     * Solve the nonlinear system of equations $F(x) = 0$.
+     *
+     * This function returns the number of iterations.
+     * The vector @p x must contain the initial guess.
+     * Upon returning, the @p x vector contains the solution.
+     *
+     * Here we also set the matrices to handle Jacobians.
+     */
+    unsigned int
+    solve(VectorType &x, AMatrixType &A, PMatrixType &P);
```

Comment:
I am wondering if it makes really sense to have all these variants. The latter two are the same as `solve` preceded by `reinit_matrix`/`reinit_matrices`? Maybe we should provide a preferred approach. Similar, we could also attach `b` optionally!?

Reply:
those are shorthand notations for users that also allow testing the interface.  For me it makes sense to keep them. @bangerth ?

Reply:
> Yes, I'm not a fan of having 6 different versions of this function.
> 
> We could already reduce this by half if we always assumed b=0. What's the use case where a user cannot just roll that vector into the residual?
> 

This was a very old design choice in PETSc. In deal.II. perspective,  you can see `SNESSolve` as a `vmult` method, thus you can support a nonzero input vector.

Reply:
I don't see the use of $b$ as particularly important, and it also does not match the interfaces that we have for KINSOL and NOX. My vote would be to just assume everyone wants to solve $F(x)=0$ and just not expose PETSc's ability to deal with a vector $b$ in our wrappers.

Reply:
Ok, will do